### PR TITLE
Sd

### PR DESCRIPTION
**Pull request by pdaures**
PR for service discovery with Consul

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -38,7 +38,9 @@ func parseProperties(reader io.Reader) map[string]string {
 
 		split := strings.Split(line, "=")
 		if len(split) < 2 {
-			log.Printf("WARN: cannot parse config line %s\n", line)
+			if len(line) > 0 {
+				log.Printf("WARN: cannot parse config line %s\n", line)
```
Comment (pdaures) [inline config.go]: :+1:

Comment (pdaures) [reply config.go]: I wonder if most if this code is useful or not, Viper has native support for Properties files.
(It's not something to fix in the scope of THIS PR)

File: main.go [hunk 2]
```diff
@@ -1,39 +1,64 @@
 package gorillaz
```
Comment (pdaures) [inline main.go]: Side note:
We should rename this file, it's confusing to have the name of something runnable in a library

File: main.go [hunk 3]
```diff
@@ -1,39 +1,64 @@
...
-	Router *mux.Router
+	Router             *mux.Router
+	ServiceDiscovery   ServiceDiscovery
+	StreamProviderPort int
 	// use int32 because sync.atomic package doesn't support boolean out of the box
 	isReady        *int32
 	isLive         *int32
 	grpcServer     *grpc.Server
 	streamRegistry *streamRegistry
+	httpListener   *net.Listener
```
Comment (pdaures) [inline main.go]: Do we need both the port and the listener?

File: main.go [hunk 4]
```diff
@@ -1,39 +1,64 @@
...
+	return func(g *Gaz) error {
+		g.context = context
+		return nil
+	}
 }
 
 // New initializes the different modules (Logger, Tracing, Metrics, ready and live Probes and Properties)
 // It takes root at the current folder for properties file and a map of properties
-func New(context map[string]interface{}) *Gaz {
+func New(options ...Option) *Gaz {
```
Comment (pdaures) [inline main.go]: thanks!
Let's not put gorillaz.New(nil) everywhere

File: main.go [hunk 5]
```diff
@@ -1,39 +1,64 @@
...
 	streamRegistry *streamRegistry
+	httpListener   *net.Listener
+	httpPort       int
+	context        map[string]interface{}
+}
+
+type Option func(*Gaz) error
+
+func WithContext(context map[string]interface{}) Option {
+	return func(g *Gaz) error {
```
Comment (pdaures) [inline main.go]: Do we really need this one?
I don't know how to use this Context anyway, passing magical map[string]interface{} is not usable for me

File: main.go (hunk 3 above)
Comment (pdaures) [reply main.go]: Listener is an interface, so you don't need a pointer to it

File: main.go [hunk 6]
```diff
@@ -63,20 +88,36 @@ func New(context map[string]interface{}) *Gaz {
 	}
 	stream.RegisterStreamServer(gaz.grpcServer, gaz.streamRegistry)
 
+	port := viper.GetInt("http.port")
+	httpListener, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		panic(err)
+	}
+	gaz.httpPort = httpListener.Addr().(*net.TCPAddr).Port
+	gaz.httpListener = &httpListener
```
Comment (pdaures) [inline main.go]: As written above, Listener is an interface, so no need for &httpListener

File: main.go [hunk 7]
```diff
@@ -95,13 +136,41 @@ func (g Gaz) Run() {
 
 		// register /version to return the build version
 		g.Router.HandleFunc("/version", VersionHTML).Methods("GET")
-
-		port := viper.GetInt("http.port")
-		Sugar.Infof("Starting http server on port %d", port)
-		err := http.ListenAndServe(fmt.Sprintf(":%d", port), g.Router)
+		Sugar.Infof("Starting HTTP server on :%d", g.httpPort)
```
Comment (pdaures) [inline main.go]: I will remove this and put it in Prometheus. But OK to keep it here for now

File: main.go [hunk 8]
```diff
@@ -95,13 +136,41 @@ func (g Gaz) Run() {
...
+		err := http.Serve(*g.httpListener, g.Router)
 		if err != nil {
-			Sugar.Errorf("Cannot start HTTP server on :%d, %+v", port, err)
+			Sugar.Errorf("Cannot updater HTTP server on :%d, %+v", g.httpPort, err)
```
Comment (pdaures) [inline main.go]: Updater?
Isn't it start?

File: main.go [hunk 9]
```diff
@@ -95,13 +136,41 @@ func (g Gaz) Run() {
...
 			panic(err)
 		}
 	}()
 }
+
+func (g Gaz) Register(d *ServiceDefinition) (string, error) {
```
Comment (pdaures) [inline main.go]: Shouldn't it be on (g *Gaz) instead of (g Gaz)?

File: main.go [hunk 10]
```diff
@@ -95,13 +136,41 @@ func (g Gaz) Run() {
...
+		PermitWithoutStream: true,
+	})
+	options := make([]grpc.DialOption, len(opts)+3)
+	options[0] = grpc.WithBalancerName(roundrobin.Name)
+	options[1] = grpc.WithBackoffMaxDelay(5 * time.Second)
+	options[2] = clientKeepAlive
+	for i, o := range opts {
+		options[3+i] = o
+	}
+	return grpc.Dial("gorillaz:///"+target, options...)
```
Comment (pdaures) [inline main.go]: oh, I see what you did here!

File: main.go [hunk 11]
```diff
@@ -95,13 +136,41 @@ func (g Gaz) Run() {
...
 			panic(err)
 		}
 	}()
 }
+
+func (g Gaz) Register(d *ServiceDefinition) (string, error) {
+	if g.ServiceDiscovery == nil {
+		return "", errors.New("no service registry configured")
+	}
+	return g.ServiceDiscovery.Register(d, g.httpPort)
```
Comment (pdaures) [inline main.go]: httpPort? not the gRPC port instead?

File: main.go [hunk 12]
```diff
@@ -95,13 +136,41 @@ func (g Gaz) Run() {
...
+}
+
+func (g Gaz) GrpcDial(target string, opts ...grpc.DialOption) (*grpc.ClientConn, error) {
+	clientKeepAlive := grpc.WithKeepaliveParams(keepalive.ClientParameters{
+		Time:                15 * time.Second,
+		PermitWithoutStream: true,
+	})
+	options := make([]grpc.DialOption, len(opts)+3)
+	options[0] = grpc.WithBalancerName(roundrobin.Name)
+	options[1] = grpc.WithBackoffMaxDelay(5 * time.Second)
```
Comment (pdaures) [inline main.go]: This should be overridable somehow

File: sd.go [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+type ServiceDefinition struct {
+	ServiceName                    string
+	Addr                           string
+	Port                           int
+	Tags                           []string
+	Meta                           map[string]string
+	Opt                            AgentServiceRegistrationOpt
+	DeregisterCriticalServiceAfter time.Duration
+	CheckInterval                  time.Duration
```
Comment (pdaures) [inline sd.go]: Please add comments to what these times mean, same way as https://github.com/grpc/grpc-go/blob/master/keepalive/keepalive.go#L37

File: sd.go [hunk 14]
```diff
@@ -0,0 +1,128 @@
...
+	"go.uber.org/zap"
+	"google.golang.org/grpc/resolver"
+	"strings"
+	"time"
+)
+
+type AgentServiceRegistrationOpt func(sd *api.AgentServiceRegistration)
+
+type ServiceDefinition struct {
+	ServiceName                    string
```
Comment (pdaures) [inline sd.go]: Do you think it makes sense to replace the definition of `Addr`, `Port`, `Tags` and `Meta` here with embedding a ServiceEndpoint?

File: main.go (hunk 11 above)
Comment (pdaures) [reply main.go]: EDIT:
I start to understand as I go down my review. For the consul health check, we need the http port. Am I correct?

File: sd.go [hunk 15]
```diff
@@ -0,0 +1,128 @@
...
+func (r *serviceDiscoveryResolver) sendUpdate() {
+	endpoints, err := r.serviceDiscovery.Resolve(r.name)
+	if err != nil {
+		Log.Warn("Error while resolving ", zap.String("name", r.name), zap.Error(err))
+		return
+	}
+	addrs := make([]resolver.Address, len(endpoints))
+	for i, e := range endpoints {
+		addrs[i] = resolver.Address{Addr: fmt.Sprintf("%s:%d", e.Addr, e.Port)}
+	}
```
Comment (pdaures) [inline sd.go]: should we sort the []addrs in the end by address or something?

File: stream_test.go [hunk 16]
```diff
@@ -49,7 +49,7 @@ func TestStreamLazy(t *testing.T) {
 
 	endpoint, err := NewStreamEndpoint(IPEndpoint, []string{addr})
 	if err != nil {
-		t.Errorf("cannot start consumer, %+v", err)
+		t.Errorf("cannot updater consumer, %+v", err)
```
Comment (pdaures) [inline stream_test.go]: I think you're going to have to fix this replace all start -> updater :D

File: sd.go (hunk 13 above)
Comment (pdaures) [reply sd.go]: EDIT: from what I see, we don't use it here.
Shouldn't these values go instead in the other piece of the puzzle, the etcd or consul part?

File: main.go (hunk 8 above)
Comment (matgabriel) [reply main.go]: probably a refactoring gone wrong :D

File: main.go (hunk 2 above)
Comment (matgabriel) [reply main.go]: yep

File: main.go (hunk 5 above)
Comment (matgabriel) [reply main.go]: I though you might be using it, but if no one uses it, then let's remove it.

File: main.go (hunk 11 above)
Comment (matgabriel) [reply main.go]: yes, but I might replace it with the gRPC health check

File: main.go (hunk 12 above)
Comment (matgabriel) [reply main.go]: yes, it the clients put these options in the config, it will override the defaults

File: main.go (hunk 3 above)
Comment (matgabriel) [reply main.go]: the gRPC health check is working, so I will remove the listener and port from the Gaz object

File: main.go (hunk 5 above)
Comment (matgabriel) [reply main.go]: I see that it is used in tests to specify a different config file

File: sd.go (hunk 15 above)
Comment (matgabriel) [reply sd.go]: what would be the point of sorting?

File: sd.go (hunk 14 above)
Comment (matgabriel) [reply sd.go]: +1

File: sd.go (hunk 13 above)
Comment (matgabriel) [reply sd.go]: I made a mistake here, I introduced a dependency on api.AgentServiceRegistration. 
The idea was instead to offer consul-agnostic options to configure service discovery

File: main.go (hunk 9 above)
Comment (matgabriel) [reply main.go]: it seems that once built, the Gaz object is not modified, so should we remove the pointer from all methods?

File: main.go (hunk 5 above)
Comment (matgabriel) [reply main.go]: this was replaced by the WithConfigPath option

**Pull request merged**